### PR TITLE
Wait for git commit status to change after creating datasets in integration tests

### PR DESCRIPTION
**Pull request by rlc2**
Wait for Conservator to commit a new dataset before operating on it for integration tests that involve datasets.

Also, when testing dataset deletion, wait for the dataset's ID to become invalid, instead of counting dataset search results.

**File Changes:**

`FLIR/conservator/wrappers/dataset.py`

- Add `wait_for_dataset_commit()` method to Dataset objects for clients or integration tests that need to wait on a new dataset's commit state.

`test/integration/conftest.py`

- Only create the database admin user once.

`test/integration/test_dataset.py`

- Remove the `@pytest.mark.skip()` decorator from `test_delete()`.
- Wait for the dataset commit state to change after every dataset creation.
- In `test_delete()`, check that the dataset ID becomes invalid after deletion, instead of relying on the total dataset count to be 0.
- In `test_add_get_frames_reversed()`, specify the image IDs used in the test, instead of relying on unfiltered image search results.

`test/integration/test_collections.py`

- Remove the `@pytest.mark.skip()` decorator from `test_download_datasets()`.
- Wait for the dataset commit state to change after every dataset creation.

`test/integration/test_cvc.py`

- Wait for the dataset commit status to change after every dataset creation.

`test/integration/test_dataset_frame.py`

- Wait for the created dataset's git commit state to change in the `init_dataset()` method.

`test/integration/test_file_locker.py`

- Wait for the created dataset's git commit state to change in `test_dataset_file_locker()`.

`test/integration/test_local_dataset.py`

- Wait for the dataset commit state to change after every dataset creation.

`test/integration/test_metadata.py`

- Wait for the created dataset's git commit state to change in `test_metadata_download_upload_for_dataset()`, in case its `@pytest.mark.skip()` decorator gets removed.

[[JIRA Task](https://jiracommercial.flir.com/browse/CON-2808)]

### CODE REVIEW

**Review by SJLC — APPROVED** [review]

**Review by nbutler77 — APPROVED** [review]
The changes look reasonable, and they seem to have passed on Jenkins, but the same test cases still fail for me locally.

File: test/integration/test_dataset.py [hunk 1]
```diff
@@ -34,13 +34,32 @@ def test_populate(conservator):
     assert dataset_from_id.name == dataset.name
 
 
-@pytest.mark.skip()
 def test_delete(conservator):
     dataset = conservator.datasets.create("My dataset")
-    assert conservator.datasets.count() == 1
+    # Give the server time to create and commit the dataset.
```
Comment (shiftycow) [inline test/integration/test_dataset.py:39]: I don't think that this section from 39-49 actually does anything useful; once the `conservator.datasets.create()` returns, it returns the dataset with an ID, which means that it is in conservator. 

The check that really needs to happen is to poll the dataset by ID and check it's commit status field

File: test/integration/test_dataset.py [hunk 2]
```diff
@@ -34,13 +34,32 @@ def test_populate(conservator):
...
+            break
+    assert found
 
     dataset.delete()
-    assert conservator.datasets.count() == 0
+    # Give the server time to commit and delete the dataset.
+    found = False
+    for _ in range(10):
+        time.sleep(1)
+        for dset in conservator.datasets.all():
```
Comment (shiftycow) [inline test/integration/test_dataset.py:56]: This could, likewise, just query for the one dataset by ID, which should return a result until the delete process is finished. 10 seconds may or may not be enough for that to complete either.

**Review by shiftycow — CHANGES_REQUESTED** [review]

File: test/integration/test_dataset.py (hunk 2 above)
Comment (rlc2) [reply test/integration/test_dataset.py:56]: The integration tests are run on a local Kind instance, so there shouldn't be any other activity slowing down the server and few, if any other datasets in the search results.  I'm surprised to hear that deleting an empty dataset under those conditions could take longer than 10 seconds.  When I ran the integration tests locally, I saw errors until I added the delays, which tells me there's a possible race condition.  Without visibility into what's happening on the server side, adding delays is the simplest method to handle those.

**Comment by rlc2** [comment]
> The changes look reasonable, and they seem to have passed on Jenkins, but the same test cases still fail for me locally.

Could you post the errors from your local tests somewhere?  Debugging these issues is difficult, since I can't reproduce them locally.

File: test/integration/test_dataset.py (hunk 1 above)
Comment (rlc2) [reply test/integration/test_dataset.py]: Added a new commit that waits on the git_commit_state field to change to "completed" before checking out or deleting the dataset.

File: test/integration/test_dataset.py (hunk 2 above)
Comment (rlc2) [reply test/integration/test_dataset.py]: Added a new commit that checks that the dataset ID becomes invalid after deleting the dataset.

**Review by nbutler77 — CHANGES_REQUESTED** [review]
Can you refactor how the test harness interacts with the database? I suspect part of the problem is that the user is deleted so often, so if we could create a user at the start of a test run, and have that same user until all tests are run, that might solve some of the issues.

In general, it looks like *all* dataset creation in the test suite should wait for the dataset to be committed, as many of them are failing in the `dataset-commit` worker. It looks like the datasets are being deleted from the database before the commit worker can process them.

File: test/integration/test_collections.py [hunk 3]
```diff
@@ -320,6 +324,7 @@ def test_get_datasets(conservator):
 def test_download_datasets(conservator, tmp_cwd):
     collection = conservator.collections.create_from_remote_path("/Some/Collection")
     dataset_1 = collection.create_dataset("My first dataset")
+    assert wait_for_dataset_commit(conservator, dataset_1.id)
     # Wait for the server to create and commit the dataset.
```
Comment (SJLC) [inline test/integration/test_collections.py:328]: Didn't you mean to replace the "Wait for the server to create and commit the dataset" section with wait_for_dataset_commit()?

Comment (rlc2) [reply test/integration/test_collections.py]: Removed redundant code.

File: test/integration/conftest.py [hunk 4]
```diff
@@ -141,11 +142,23 @@ def db(mongo_client):
...
+    admin_found = False
+    admin_email = "admin@example.com"
+    if "TEST_ADMIN_EMAIL" in os.environ:
+        admin_email = os.environ["TEST_ADMIN_EMAIL"]
     for name in db.list_collection_names():
         if name.startswith("system."):
             continue
         if name in PRESERVED_COLLECTIONS:
             continue
+        if name == "users":
```
Comment (nbutler77) [inline test/integration/conftest.py:154]: I don't understand why you "special case" the `users` collection here. Is there any reason you can't explicitly check whether the admin user is already present (e.g. `db.users.find({email: admin_email})`) and insert it if it isn't present before this loop, and then just add `users` to `PRESERVED_COLLECTIONS`?

File: test/integration/conftest.py [hunk 5]
```diff
@@ -246,3 +267,16 @@ def upload_media(conservator, media):
         )
         media_ids.append(media_id)
     conservator.media.wait_for_processing(media_ids)
+
+
+def wait_for_dataset_commit(cns, dataset_id):
```
Comment (nbutler77) [inline test/integration/conftest.py:272]: Just an observation - it might be useful to have this function in the main codebase, possibly in `local_dataset`.

**Review by nbutler77 — CHANGES_REQUESTED** [review]
Integration tests now run as expected locally; some code changes requested

File: test/integration/conftest.py (hunk 4 above)
Comment (rlc2) [reply test/integration/conftest.py:154]: I left the user account creation in the `conservator()` fixture (defined just after `empty_db()`), since it needs to collect the API key to return a `Conservator` object there anyhow.

Comment (nbutler77) [reply test/integration/conftest.py:154]: I didn't notice it was in a separate function. In any case:
 - There is no reason to clear out the `users` collection - I don't think there are any integration tests that rely on it being empty, and it's an annoyance on developer machines if you're trying to log in using a pre-created admin user
 - The way the admin user is searched for and created in the `conservator()` fixture seems sub-optimal. The logic, IMHO, should be:
     - Search for user by email address
         - If it's there, check that it has the expected API key
             - If it does, break
             - If it doesn't, update the record to have the expected API key
         - If it's not there, create it

Comment (rlc2) [reply test/integration/conftest.py:154]: Also, I discovered that there was already an account with the default `admin@example.com` once the Kind cluster is up and running, but it has no `apiKey` field in it, so it can be a bit more complicated than the `db.users.find({email: admin_email})` query.

File: test/integration/conftest.py (hunk 5 above)
Comment (rlc2) [reply test/integration/conftest.py:272]: Would it be worth adding a call to this method inside `Dataset.commit()`, just to avoid the possibility that user code triggers the race condition by creating a dataset then quickly deleting it again?

Comment (nbutler77) [reply test/integration/conftest.py:272]: I don't think so. A user can do stuff with datasets immediately without worrying about commits (e.g. adding frames, changing permissions, etc.) - I think a user creating a dataset and immediately deleting it is an edge case, and not worth delaying creation of a dataset over.

File: test/integration/conftest.py [hunk 6]
```diff
@@ -140,7 +141,11 @@ def db(mongo_client):
 
 @pytest.fixture(scope="class")
 def empty_db(db):
-    PRESERVED_COLLECTIONS = ["groups", "organizations", "allowedDomains"]
+    PRESERVED_COLLECTIONS = ["users", "groups", "organizations", "allowedDomains"]
+    admin_found = False
```
Comment (nbutler77) [inline test/integration/conftest.py:145]: `admin_found` and `admin_email` are redundant here

File: test/integration/conftest.py [hunk 7]
```diff
@@ -171,16 +176,28 @@ def conservator(empty_db):
...
-        {
-            "_id": Conservator.generate_id(),
-            "role": ADMIN_ROLE,
-            "name": "admin user",
-            "email": admin_email,
-            "apiKey": api_key,
-            "organizationId": organization["_id"],
-        }
-    )
+    admin_found = False
```
Comment (nbutler77) [inline test/integration/conftest.py:179]: This variable seems redundant to me. The `if` statement below could be rewritten as:
```
if user: 
...
else:
    # create user
```

File: test/integration/conftest.py (hunk 5 above)
Comment (nbutler77) [reply test/integration/conftest.py:268]: Can you move this function to the main codebase, please?

**Review by nbutler77 — CHANGES_REQUESTED** [review]

File: test/integration/conftest.py (hunk 7 above)
Comment (rlc2) [reply test/integration/conftest.py]: Done.

File: test/integration/conftest.py (hunk 6 above)
Comment (rlc2) [reply test/integration/conftest.py]: Removed.

File: test/integration/conftest.py (hunk 5 above)
Comment (rlc2) [reply test/integration/conftest.py]: Moved into dataset.py.

File: test/integration/conftest.py (hunk 4 above)
Comment (rlc2) [reply test/integration/conftest.py]: Implemented the proposed logic change for the conservator() fixture.

**Review by nbutler77 — APPROVED** [review]
Code changes made as requested, all tests seem to pass on Jenkins.
